### PR TITLE
HIVE-24992: Incremental rebuild of MV having aggregate in presence of delete operation

### PR DESCRIPTION
**Pull request by kasakrisz**
### What changes were proposed in this pull request?
* Introduce a new materialized view rebuild rewrite rule: transform insert overwrite plan to incremental rebuild plan in when MV definition has aggregate and delete operations present since the last rebuild.
* Above rule can be used only if the definition contains a count(*) aggregation so this extra check is added to `MaterializedViewRewritingRelVisitor`
* Extract common parts from `HiveAggregate*IncrementalRewritingRule` claases and CalcitePlanner. fixUpAST*IncrementalRebuild methods

### Why are the changes needed?
Extension of #2119

### Does this PR introduce _any_ user-facing change?
Yes. Plan of Materialized view rebuild may change when user executes commands like `EXPLAIN [CBO] ALTER MATERIALIZED VIEW <mv name> REBUILD;`

### How was this patch tested?
```
mvn test -Dtest.output.overwrite -DskipSparkTests -Dtest=TestMiniLlapLocalCliDriver -Dqfile=materialized_view_create_rewrite_4.q,materialized_view_create_rewrite_nulls.q,materialized_view_create_rewrite_one_key_gby.q,materialized_view_create_rewrite_6.q,materialized_view_create_rewrite_5.q -pl itests/qtest -Pitests
```

### CODE REVIEW

File: ql/src/test/queries/clientpositive/materialized_view_create_rewrite_6.q [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+insert into t2(a, b) values
+('update', 10),
+('null_update', null);
+insert into t2(a, b) values ('remove', 30), ('null_remove', null);
+insert into t2(a, b) values ('sum0', 0);
+
+-- Aggregate with count(*): incremental rebuild should be triggered even if there were deletes from source table
+create materialized view mat1 stored as orc TBLPROPERTIES ('transactional'='true') as
+select t1.a, sum(t1.b), count(*) from t1
```
Comment (jcamachor) [inline ql/src/test/queries/clientpositive/materialized_view_create_rewrite_6.q:22]: Can we add the corresponding test without `count(*)`, making sure that we fallback to full rebuild?

**Review by jcamachor — COMMENTED** [review]
A couple of comments.

We should add information about whether a materialized view supports incremental review or not in an additional column in `show materialized views` statement (see https://github.com/apache/hive/blob/master/ql/src/test/results/clientpositive/llap/show_materialized_views.q.out for reference). That would be quite useful for users to have the correct expectation. Can we create a follow-up for that and tackle it?

In addition, could we update the MV wiki in Hive with the latest improvements (also follow-up)? https://cwiki.apache.org/confluence/display/Hive/Materialized+views

Thanks

File: ql/src/test/queries/clientpositive/materialized_view_create_rewrite_6.q [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+join t2 on (t1.a = t2.a)
+group by t1.a;
+
+
+-- rebuild the view (incrementally)
+explain cbo
+alter materialized view mat1 rebuild;
+explain
+alter materialized view mat1 rebuild;
+alter materialized view mat1 rebuild;
```
Comment (jcamachor) [inline ql/src/test/queries/clientpositive/materialized_view_create_rewrite_6.q:59]: Does this example contain rows that hit the three branches in the merge, i.e., insert, update, delete?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAggregateIncrementalRewritingRuleBase.java [hunk 3]
```diff
@@ -203,17 +147,42 @@ public void onMatch(RelOptRuleCall call) {
...
-        .filter(filterCond)
+        .filter(createFilterCondition(joinRightInput, flagNode, projExprs, relBuilder))
         .project(projExprs)
         .build();
     call.transformTo(newNode);
   }
+
+  protected abstract T createJoinRightInput(RelOptRuleCall call);
+
+  protected static class RightInput {
```
[the remark this answers is not in the document]
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAggregateIncrementalRewritingRuleBase.java]: Renamed

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAggregateInsertDeleteIncrementalRewritingRule.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+ *   FROM TAB_A
+ *   JOIN TAB_B ON (TAB_A.a = TAB_B.z)
+ *   WHERE TAB_A.ROW_ID &gt; 5
+ *   GROUP BY a, b) source
+ * ON (mv.a <=> source.a AND mv.b <=> source.b)
+ * WHEN MATCHED AND mv.c + source.c &lt;&gt; 0
+ *   THEN UPDATE SET mv.s = mv.s + source.s, mv.c = mv.c + source.c
+ * WHEN NOT MATCHED
+ *   THEN INSERT VALUES (source.a, source.b, s, c);
+ * WHEN MATCHED AND countStar = 0 THEN DELETE
```
[the remark this answers is not in the document]
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAggregateInsertDeleteIncrementalRewritingRule.java]: fixed

File: ql/src/test/queries/clientpositive/materialized_view_create_rewrite_6.q (hunk 1 above)
Comment (kasakrisz) [reply ql/src/test/queries/clientpositive/materialized_view_create_rewrite_6.q:22]: `materialized_view_create_rewrite_4.q` covers that case.

File: ql/src/test/queries/clientpositive/materialized_view_create_rewrite_6.q (hunk 2 above)
Comment (kasakrisz) [reply ql/src/test/queries/clientpositive/materialized_view_create_rewrite_6.q:65]: I found that I inserted rows to only one of the source tables that is why new records were not inserted into the view. Added some inserts to the second source table. All three branches are covered.

**Comment by kasakrisz** [comment]
Created a follow-up for extending 'show materialized views': https://issues.apache.org/jira/browse/HIVE-25066

**Review by jcamachor — APPROVED** [review]

**Pull request merged**